infra/batch/datasets/main.py

```diff
@@ -0,0 +1,225 @@
+    :param engine: Database engine
+    :param url: Producer URL
+    :param bucket_name: GCP bucket name
+    :param stable_id: Feed's stable ID
+    :param feed_id: Feed's ID
+    """
+    transaction = None
+    connection = None
+    errors = ""
+    try:
+        md5_file_hash, hosted_url = upload_dataset(url, bucket_name, stable_id)
+
+        # Set up transaction for SQL updates
+        connection = engine.connect()
+        transaction = connection.begin()
+
+        # Create a new version of the dataset in the database
+        select_dataset_statement = text(f"select id, hash from gtfsdataset where latest=true and feed_id='{feed_id}'")
+        dataset_results = connection.execute(select_dataset_statement).all()
+        dataset_id = dataset_results[0][0] if len(dataset_results) > 0 else None
+        dataset_hash = dataset_results[0][1] if len(dataset_results) > 0 else None
+        print(f"Dataset ID = {dataset_id}, Dataset Hash = {dataset_hash}")
+        if dataset_id is None:
+            errors += f"[INTERNAL ERROR] Couldn't find latest dataset related to feed_id {feed_id}\n"
+            return
+
+        # Set the previous version latest field to false
+        if dataset_hash is not None and dataset_hash != md5_file_hash:
+            sql_statement = f"update gtfsdataset set latest=false where id='{dataset_id}'"
+            connection.execute(text(sql_statement))
+
+        sql_statement = f"insert into gtfsdataset (id, feed_id, latest, bounding_box, note, hash, " \
+                        f"download_date, stable_id, hosted_url) " \
+                        f"select '{str(uuid.uuid4())}', feed_id, true, bounding_box, note, " \
+                        f"'{md5_file_hash}', NOW(), stable_id, '{hosted_url}' from " \
+                        f"gtfsdataset where id='{dataset_id}'"
+
+        # In case the dataset doesn't include a hash or the dataset was deleted from the bucket,
+        # update the existing entity
+        if dataset_hash is None or dataset_hash == md5_file_hash:
+            sql_statement = f"update gtfsdataset set hash='{md5_file_hash}', hosted_url='{hosted_url}' where id='{dataset_id}'"
+        connection.execute(text(sql_statement))
+
+        # Commit transaction after every step has run successfully
+        transaction.commit()
+    except Exception as e:
+        if transaction is not None:
+            transaction.rollback()
+        error_traceback = traceback.format_exc()
+        errors += f"[ERROR]: {e}\n{error_traceback}\n"
+        print(f"Logging errors for stable id {stable_id}\n{errors}")
+        storage_client = storage.Client()
+        bucket = storage_client.get_bucket(bucket_name)
+        error_type = 'other'
+        if 'sqlalchemy' in errors:
+            error_type = 'sqlachemy'
+        elif isinstance(e, HTTPError):
+            error_type = f"http/{e.response.status_code}"
+        blob = bucket.blob(f"errors/{datetime.now().strftime('%Y%m%d')}/{error_type}/{stable_id}.log")
```

Comment:
Having the logs in a bucket is hard to consume without an additional orchestration or tool to query the log files. I suggest removing this section here and printing the logs to the console for MVP. This will actually send the logs to `Google Cloud's operations suite` which can be consumed and queried using the `GCP Logs Explorer`. We should explore how we can format the logs in a way that queries on Logs Explorer are easy to create, in any case this can be part of different PR/issue.

Reply:
Created [this issue](https://github.com/orgs/MobilityData/projects/64/views/1?pane=issue&itemId=39289533) to address the problem.